bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerCreateOp.java

```diff
@@ -200,4 +209,233 @@ private void createComplete(int rc, LedgerHandle lh) {
+                return false;
+            }
+
+            if (builderCustomMetadata == null) {
+                LOG.error("invalid null customMetadata");
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            return true;
+        }
+
+        @Override
+        public CompletableFuture<WriteHandle> execute() {
+            CompletableFuture<WriteHandle> counter = new CompletableFuture<>();
+            create(new SyncCreateCallback(), counter);
+            return counter;
+        }
+
+        private void create(CreateCallback cb, Object ctx) {
+            if (!validate(cb, ctx)) {
+                return;
+            }
+            LedgerCreateOp op = new LedgerCreateOp(bk, builderEnsembleSize,
+                builderWriteQuorumSize, builderAckQuorumSize, DigestType.fromApiDigestType(builderDigestType),
+                builderPassword, cb, ctx, builderCustomMetadata);
+            ReentrantReadWriteLock closeLock = bk.getCloseLock();
+            closeLock.readLock().lock();
+            try {
+                if (bk.isClosed()) {
+                    cb.createComplete(BKException.Code.ClientClosedException, null, ctx);
+                    return;
+                }
+                op.initiate();
+            } finally {
+                closeLock.readLock().unlock();
+            }
+        }
+    }
+
+    private static class CreateAdvBuilderImpl implements CreateAdvBuilder {
+
+        private Long builderLedgerId;
+        private final int builderEnsembleSize;
+        private final int builderAckQuorumSize;
+        private final int builderWriteQuorumSize;
+        private final byte[] builderPassword;
+        private final Map<String, byte[]> builderCustomMetadata;
+        private final org.apache.bookkeeper.client.api.DigestType builderDigestType;
+        private final BookKeeper bk;
+
+         private CreateAdvBuilderImpl(CreateBuilderImpl other) {
+            this.builderEnsembleSize = other.builderEnsembleSize;
+            this.builderAckQuorumSize = other.builderAckQuorumSize;
+            this.builderWriteQuorumSize = other.builderWriteQuorumSize;
+            this.builderPassword = other.builderPassword;
+            this.builderDigestType = other.builderDigestType;
+            this.builderCustomMetadata = other.builderCustomMetadata;
+            this.bk = other.bk;
+        }
+
+        @Override
+        public CreateAdvBuilder withLedgerId(long ledgerId) {
+            builderLedgerId = ledgerId;
+            return this;
+        }
+
+        @Override
+        public CompletableFuture<WriteAdvHandle> execute() {
+            CompletableFuture<WriteAdvHandle> counter = new CompletableFuture<>();
+            create(new SyncCreateCallback(), counter);
+            return counter;
+        }
+
+        private boolean validateAdv(CreateCallback cb, Object ctx) {
```

Comment:
there are duplicated logics as CreateBuilderImpl. if you keep a reference of `CreateBuilderImpl`, you can then use `CreateBuilderImpl#validate` to validate the common parts.